Python/chapter01/1.7 - Rotate Matrix/miguel_1.7_sol.py

```diff
@@ -0,0 +1,174 @@
+"""
+Python version 3.7.0
+1.7 - Rotate Matrix
+Given an image represented by an NxN matrix, where each pixel in the image is 4 bytes, write a method to rotate
+the image by 90 degrees.  Can you do this in place?
+"""
+import unittest
+from typing import List
+
+
+def rotate_matrix(matrix: List[List[int]], N: int) -> List[List[int]]:
+    """
+    Rotate matrix will rotate the given matrix by 90 degrees.
+    Runtime: O(N^2), asymptotic runtime depends on N
+    Space Complexity: O(N^2), creating a new matrix of NxN called 'rotated'
+    :param matrix: an NxN matrix
+    :param N: the size of the matrix (NxN)
+    :return: a newly rotated matrix
+    """
+    rotated = [[0 for i in range(N)] for j in range(N)]
+    for i in range(N):
+        for j in range(N):
+            rotated[j][(N-1)-i] = matrix[i][j]
+    return rotated
+
+
+def rotate_matrix_in_place(matrix: List[List[int]], start_row: int, start_col: int, N: int) -> List[List[int]]:
+    """
+    Does the same as rotate_matrix, but in place.
+    Runtime: O(N^2), asymptotic runtime depends on N. We make N^2 swaps.
+    Space Complexity: O(1), constant amount of temp variables that does not depend on N.
+                        Note: I am assuming that compiler will optimize the tail recursion.
+    :param matrix: an NxN matrix
+    :param start_row: starting row index
+    :param start_col: starting col index
+    :param N: the size of the matrix (NxN)
+    :return: the input matrix, but rotated
+    """
+    num_rotations = 4
+    if N == 0 or N == 1:
+        return matrix
+    col = start_col
+    while True:
+        rotated_row = start_row
+        rotated_col = col
+        temp_new = matrix[start_row][col]
+        for r in range(0, num_rotations):
+            temp = temp_new
+            # compute new rotated indices
+            prev_col = rotated_col
+            rotated_col = N - 1 - rotated_row + (start_row * 2)  # offset to account for reduced N
+            rotated_row = prev_col
+            # store value at newly computed indices
+            temp_new = matrix[rotated_row][rotated_col]
+            matrix[rotated_row][rotated_col] = temp
+        if col - start_col >= N - 2:
+            break
+        col = col + 1
+    return rotate_matrix_in_place(matrix, start_row + 1, start_col + 1, N - 2)
```

Comment:
same here: you are mixing two different styles of looping, and both can be easily replaced with a regular for loop.

one thing you _can_ do is extract part of the functionality of this function to a helper function, if you find that to be cleaner.

Reply:
Extracted the rotations to a helper function.  Does it follow convention to prepend this new helper function with '_' since it is used internally?

Reply:
it does!